File: src/agentsfleetd/fleet/assign.zig
//! Lease assignment — choose the next fleet + event for a polling runner.
//!
//! One pass per `lease` call (no server-side long-poll loop; the runner
//! re-polls via `retry_after_ms`). The pass is **ready-first**:
//!
//!   0. `fleet_ready.peek` — a bounded, randomized read of the shared readiness
//!      index, BEFORE any pool acquisition. An empty index answers no-work with
//!      zero Postgres round-trips, which is the dominant steady state on any
//!      deployment holding more fleets than concurrent events.
//!   1. The candidate query, restricted to those fleets and capped at
//!      `MAX_READY_CANDIDATES_PER_POLL`. Readiness narrows the input; it never
//!      decides eligibility — the label gate and sticky ordering are properties
//!      of the query and stay exactly where they were.
//!
//! Then, per candidate:
//!
//!   2. `affinity.claim` — the atomic per-fleet CLAIM. Exactly one of N racing
//!      runners wins the slot; a loser gets `.taken` and moves on, having read
//!      no event (claim precedes the read ⇒ nothing is orphaned).
//!   3. won + a prior `active` lease exists  → RECLAIM that dead holder's event
//!      from Postgres (no Redis re-read, no re-billing).
//!   4. won + no prior active lease           → FRESH: non-blocking XREADGROUP;
//!      no event ⇒ release the claim, clear readiness, try the next candidate.
//!
//! Every non-success exit after a win frees the claim — the no-work branches
//! inline, the post-claim error branches via `releaseWonClaim`. A transient
//! reclaim or envelope-allocation failure therefore costs one poll, not a full
//! `LEASE_TTL_MS` stall on that fleet.
//!
//! A run of failing Redis reads ends the candidate loop early rather than
//! holding the pooled connection through one timeout per remaining candidate —
//! see `PollCost.zig`.
//!
//! The result envelope is arena-dup'd into `select`'s `alloc` (the request arena
//! in production); the caller (service.zig) loads the session + bills (fresh) or
//! reuses billing (reclaim) + issues.

const std = @import("std");
const pg = @import("pg");
const logging = @import("log");
const ec = @import("../errors/error_registry.zig");
const PgQuery = @import("../db/pg_query.zig").PgQuery;

const handlers_common = @import("../http/handlers/common.zig");
const affinity = @import("affinity.zig");
const reclaim = @import("reclaim.zig");
const sql = @import("sql.zig");
const constants = @import("common");
const redis_fleet = @import("../queue/redis_fleet.zig");
const fleet_ready = @import("../queue/fleet_ready.zig");
const queue_redis = @import("../queue/redis_client.zig");
const fleet_config = @import("../fleet_runtime/config.zig");
const PollCost = @import("PollCost.zig");

const Context = handlers_common.Context;
const log = logging.scoped(.runner_assign);

pub const Kind = enum { fresh, reclaim };

/// Billing context carried forward on a reclaim (reused, never re-charged).
pub const Reused = struct {
    tenant_id: []const u8,
    posture: []const u8,
    model: []const u8,
};

/// The chosen work: the claimed fleet + fencing token + event envelope. For a
/// reclaim, `reused` carries the prior lease's billing; for fresh it is null
/// and the caller bills. All slices arena-dup'd.
pub const Acquired = struct {
    fleet_id: []const u8,
    fencing_token: u64,
    leased_until: i64,
    kind: Kind,
    event_id: []const u8,
    actor: []const u8,
    event_type: []const u8,
    request_json: []const u8,
    workspace_id: []const u8,
    event_created_at: i64,
    reused: ?Reused = null,
};

/// A ready fleet that survived the candidate query, carrying the readiness token
/// observed when it was peeked. The token travels to the clear site: readiness may
/// only be removed for this fleet if the stored token is still this one.
///
/// Null only if the peeked id and the query's `id::text` failed to match as text,
/// which canonical lowercase UUIDs make impossible — a null then costs a skipped
/// clear (the sweeper recovers) rather than an unsafe unconditional delete.
const Candidate = struct {
    fleet_id: []const u8,
    ready_token: ?[]const u8,
};

/// Select the next work for `runner_id`, or null when nothing is leasable this
/// pass. Errors are logged and collapse to null (the runner backs off + re-polls).
///
/// `alloc` is the request arena every returned slice is dup'd into. It is passed
/// explicitly rather than read off a handler context so a test can inject an
/// allocation failure and prove the won claim is released on that exit.
pub fn select(ctx: *Context, alloc: std.mem.Allocator, runner_id: []const u8) ?Acquired {
    var cost = PollCost{};
    defer cost.report();
    return selectInner(ctx, alloc, runner_id, &cost) catch |err| {
        log.warn("assign_failed", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .runner_id = runner_id, .err = @errorName(err) });
        return null;
    };
}

fn selectInner(ctx: *Context, alloc: std.mem.Allocator, runner_id: []const u8, cost: *PollCost) !?Acquired {
    // A failed peek is retryable, never fatal, and never falls back to an
    // unbounded scan: answer no-work with the existing backoff hint and let the
    // runner re-poll (RULE ECL).
    const ready = fleet_ready.peek(ctx.queue, alloc, constants.MAX_READY_CANDIDATES_PER_POLL) catch |err| {
        log.warn("assign_ready_peek_failed", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .runner_id = runner_id, .err = @errorName(err) });
        return null;
    };
    defer fleet_ready.freePeeked(alloc, ready);

    // The zero-Postgres path. Returning here — above `pool.acquire()` — is what
    // makes idle cost scale with runner count alone instead of runners × fleets.
    if (ready.len == 0) return null;

    const conn = try ctx.pool.acquire();
    defer ctx.pool.release(conn);
    const candidates = try listReadyCandidates(conn, alloc, runner_id, ready, cost);
    for (candidates) |candidate| {
        cost.candidates_examined += 1;
        if (try tryCandidate(ctx, conn, alloc, runner_id, candidate, cost)) |acq| return acq;
        // Stop rather than pin `conn` through one Redis timeout per remaining
        // candidate; no-work is what the exhausted loop answers anyway.
        if (cost.redisBrownedOut()) {
            log.warn("assign_redis_brownout_bailout", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .runner_id = runner_id, .candidates_examined = cost.candidates_examined });
            break;
        }
    }
    return null;
}

/// Run the bounded candidate query over the peeked fleets and re-attach each
/// survivor's readiness token, preserving the SQL's ordering.
///
/// The ordering must come from the query and not from the peek, because sticky
/// preference lives in its ORDER BY. The token lookup is linear over at most the
/// per-poll ceiling of entries, which is cheaper than building a map.
fn listReadyCandidates(
    conn: *pg.Conn,
    alloc: std.mem.Allocator,
    runner_id: []const u8,
    ready: []const fleet_ready.Ready,
    cost: *PollCost,
) ![]Candidate {
    const ids = try alloc.alloc([]const u8, ready.len);
    for (ready, 0..) |entry, i| ids[i] = entry.fleet_id;

    cost.countDb(1);
    var q = PgQuery.from(try conn.query(sql.SELECT_READY_CANDIDATES, .{
        fleet_config.FleetStatus.active.toSlice(),
        runner_id,
        ids,
        @as(i64, @intCast(constants.MAX_READY_CANDIDATES_PER_POLL)),
    }));
    defer q.deinit();

    var out: std.ArrayList(Candidate) = .empty;
    while (try q.next()) |row| {
        const fleet_id = try alloc.dupe(u8, try row.get([]const u8, 0));
        try out.append(alloc, .{ .fleet_id = fleet_id, .ready_token = tokenFor(ready, fleet_id) });
    }
    return out.toOwnedSlice(alloc);
}

fn tokenFor(ready: []const fleet_ready.Ready, fleet_id: []const u8) ?[]const u8 {
    for (ready) |entry| {
        if (std.mem.eql(u8, entry.fleet_id, fleet_id)) return entry.token;
    }
    log.warn("assign_ready_token_unmatched", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .fleet_id = fleet_id });
    return null;
}

/// Free the just-won slot after a post-claim failure so the fleet is claimable
/// on the next poll instead of stalling until the claim's own LEASE_TTL_MS
/// expiry. `release` is token-guarded and idempotent, and its own failure only
/// degrades to that expiry — so it is reported, never allowed to mask the error
/// the caller is about to re-raise.
fn releaseWonClaim(conn: *pg.Conn, fleet_id: []const u8, token: u64, stage: Kind, err: anyerror) void {
    const released = if (affinity.release(conn, fleet_id, token)) |_| true else |_| false;
    log.warn("post_claim_error_released", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .fleet_id = fleet_id, .fencing_token = token, .stage = @tagName(stage), .released = released, .err = @errorName(err) });
}

/// Claim the fleet; on a win, reclaim a dead holder's event or take a fresh
/// one. Returns null when the slot is taken or has no leasable work.
fn tryCandidate(ctx: *Context, conn: *pg.Conn, alloc: std.mem.Allocator, runner_id: []const u8, candidate: Candidate, cost: *PollCost) !?Acquired {
    cost.countDb(1);
    const won = switch (try affinity.claim(conn, alloc, candidate.fleet_id, runner_id, constants.LEASE_TTL_MS)) {
        .taken => return null,
        .won => |w| w,
    };
    cost.countDb(1);
    const maybe_prior = reclaim.reclaimPriorActive(conn, alloc, candidate.fleet_id) catch |err| {
        cost.countDb(1);
        releaseWonClaim(conn, candidate.fleet_id, won.token, .reclaim, err);
        return err;
    };
    if (maybe_prior) |prior| {
        log.debug("lease_reclaimed", .{ .fleet_id = candidate.fleet_id, .event_id = prior.event_id, .lease_id = prior.lease_id, .fencing_token = won.token, .runner_id = runner_id });
        return fromReclaim(candidate.fleet_id, won, prior);
    }
    return acquireFresh(ctx, conn, alloc, candidate, won, cost);
}

/// Pull the next event for the claimed fleet: the stable consumer's own PEL
/// first (re-delivering pending-gate re-polls and sweep-recovered strands —
/// safe because this claim win proves no live lease exists), then a fresh
/// undelivered entry. No event ⇒ release the claim so the next event (and
/// other runners) are not blocked, and return null.
fn acquireFresh(ctx: *Context, conn: *pg.Conn, alloc: std.mem.Allocator, candidate: Candidate, won: affinity.Won, cost: *PollCost) !?Acquired {
    const fleet_id = candidate.fleet_id;
    // No group-ensure here. The group is created on the fleet's write path and is
    // durable, so asserting it per candidate per poll bought one Redis round-trip
    // apiece to re-learn something already true. If it is genuinely gone, the read
    // below answers `NOGROUP` and repairs itself — see `redis_fleet.readGroup`.
    var consumer_buf: [queue_redis.CONSUMER_ID_BUF_LEN]u8 = undefined;
    const consumer_id = queue_redis.stableConsumerId(&consumer_buf);
    // A failed PEL read cannot prove the PEL is empty, so it must NOT fall
    // through to the fresh read — promoting a new entry over a possibly-pending
    // gate re-poll would break own-PEL-first ordering exactly when Redis is
    // degraded. Release the claim and deliver nothing; the next poll retries
    // (consistent with the fresh-read error path below). Readiness is NOT
    // cleared on either error path, for the same reason: a failed read is not
    // evidence of an empty stream.
    var maybe_event = redis_fleet.xreadgroupFleetPending(ctx.queue, fleet_id, consumer_id) catch |err| {
        log.warn("assign_pel_read_failed", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .fleet_id = fleet_id, .err = @errorName(err) });
        cost.countDb(1);
        cost.noteRedisFailure();
        try affinity.release(conn, fleet_id, won.token);
        return null;
    };
    if (maybe_event) |ev| {
        log.debug("assign_pel_redelivered", .{ .fleet_id = fleet_id, .event_id = ev.event_id });
    } else {
        maybe_event = redis_fleet.xreadgroupFleetOnce(ctx.queue, fleet_id, consumer_id) catch |err| {
            log.warn("assign_xreadgroup_failed", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .fleet_id = fleet_id, .err = @errorName(err) });
            cost.countDb(1);
            cost.noteRedisFailure();
            try affinity.release(conn, fleet_id, won.token);
            return null;
        };
    }
    // Both arms above reached a verdict, so Redis answered — including the
    // no-event case, which is a successful read of an empty stream.
    cost.noteRedisReachable();
    var event = maybe_event orelse {
        clearReadiness(ctx, candidate);
        cost.countDb(1);
        try affinity.release(conn, fleet_id, won.token);
        return null;
    };
    defer event.deinit(ctx.queue.alloc);
    return fromFresh(alloc, fleet_id, won, &event) catch |err| {
        cost.countDb(1);
        releaseWonClaim(conn, fleet_id, won.token, .fresh, err);
        return err;
    };
}

/// The ONE place readiness is cleared, reached only where BOTH reads returned
/// null — which is the only evidence this code ever has that a fleet holds
/// nothing deliverable. Do not move this to a stream-emptiness check: ingress
/// trims at `MAXLEN ~ 10000`, so delivered entries persist and a stream is
/// essentially never empty, which would leave every fleet that ever received an
/// event permanently in the index.
///
/// Token-guarded, and that guard is load-bearing rather than defensive. The
/// affinity claim serializes runners against each other, but ingress takes no
/// claim at all: it can append and mark at any instant, including between the two
/// null reads above and this call. An unconditional delete would erase the mark
/// for that genuinely undelivered event, and nothing would rediscover it until a
/// sweep pass — turning a sub-second pickup into a multi-second one on a
/// completely healthy system.
fn clearReadiness(ctx: *Context, candidate: Candidate) void {
    const token = candidate.ready_token orelse return;
    fleet_ready.clear(ctx.queue, candidate.fleet_id, token);
}

fn fromFresh(alloc: std.mem.Allocator, fleet_id: []const u8, won: affinity.Won, event: *const redis_fleet.FleetEvent) !Acquired {
    return .{
        .fleet_id = fleet_id,
        .fencing_token = won.token,
        .leased_until = won.leased_until,
        .kind = .fresh,
        .event_id = try alloc.dupe(u8, event.event_id),
        .actor = try alloc.dupe(u8, event.actor),
        .event_type = try alloc.dupe(u8, event.event_type),
        .request_json = try alloc.dupe(u8, event.request_json),
        .workspace_id = try alloc.dupe(u8, event.workspace_id),
        .event_created_at = event.created_at_ms,
        .reused = null,
    };
}

fn fromReclaim(fleet_id: []const u8, won: affinity.Won, prior: reclaim.PriorLease) Acquired {
    return .{
        .fleet_id = fleet_id,
        .fencing_token = won.token,
        .leased_until = won.leased_until,
        .kind = .reclaim,
        .event_id = prior.event_id,
        .actor = prior.actor,
        .event_type = prior.event_type,
        .request_json = prior.request_json,
        .workspace_id = prior.workspace_id,
        .event_created_at = prior.event_created_at,
        .reused = .{ .tenant_id = prior.tenant_id, .posture = prior.posture, .model = prior.model },
    };
}

File: rustd/crates/afd_fleet/src/lease/assign.rs
//! Choosing the next fleet and event for a polling runner.
//!
//! One pass per `lease` call — no server-side long-poll loop; the runner
//! re-polls on the backoff the reply carries. The pass is READY-FIRST, and the
//! ordering is the design:
//!
//! 1. Peek the shared readiness index, BEFORE touching Postgres. An empty index
//!    answers no-work with zero database round-trips, which is the dominant
//!    steady state on any deployment holding more fleets than concurrent
//!    events.
//! 2. Run the candidate query, restricted to those fleets and capped. Readiness
//!    NARROWS the input; it never decides eligibility — the label gate and the
//!    sticky ordering are properties of the query.
//! 3. Per candidate: claim it. A loser moves on having read no event, because
//!    the claim precedes the read.
//! 4. Won with a prior active lease → RECLAIM that dead holder's event. Won with
//!    none → FRESH: the consumer's own pending list first, then a new entry.
//!
//! Every non-success exit after a win frees the claim, so an abandoned claim
//! costs one poll rather than a full TTL of silence on that fleet.

use afd_core::clock::UnixMillis;
use afd_core::id::Uuid7;
use afd_core::timing::LEASE_TTL_MS;
use afd_observability::metrics::label::fleet::RunStart;
use afd_observability::producers;
use sqlx::Row as _;

use crate::error::{Result, query};
use crate::lease::envelope::{Acquired, Kind, from_fresh, from_reclaim};
use crate::lease::sql;
use crate::lease::store::Leases;

mod diagnostics;

pub(crate) use diagnostics::warn_queue_fleet;
use diagnostics::{EVENT_LEASE_RECLAIMED, EVENT_READY_PEEK_FAILED, drop_undecodable, warn_queue};

/// Statement name, for the context a query failure carries.
const CONTEXT_CANDIDATES: &str = "lease candidate scan";

/// The `core.fleets.status` value a leasable fleet carries.
pub(crate) const FLEET_STATUS_ACTIVE: &str = "active";

/// How many ready fleets one poll will consider.
///
/// `constants.zig`'s `MAX_READY_CANDIDATES_PER_POLL`. The ceiling is what makes
/// per-poll cost independent of how many fleets exist — without it a runner
/// polling an idle deployment pays for every fleet on it, every second.
pub const MAX_READY_CANDIDATES_PER_POLL: usize = 64;

/// What one lease poll cost, gathered as it runs.
///
/// The ratio is what an operator reads: candidates per poll says how much a
/// poll examined, and round-trips per poll says how much of that reached
/// Postgres. Either number alone is unreadable, which is why they are tallied
/// together and recorded together.
#[derive(Debug, Default)]
struct PollCost {
    /// Fleets the readiness index offered this poll.
    candidates_scanned: u64,
    /// Statements this poll issued.
    database_roundtrips: u64,
}

impl Leases {
    /// Select the next work for `runner_id`, or `None` when nothing is leasable
    /// this pass.
    ///
    /// # Errors
    /// Reports a datastore that would not answer. "Nothing to do" is
    /// `Ok(None)`, not an error — the runner backs off and re-polls.
    pub async fn select(&self, runner_id: &Uuid7, now: UnixMillis) -> Result<Option<Acquired>> {
        self.select_recording(runner_id, now).await.0
    }

    /// The poll, its outcome, and what it cost.
    ///
    /// The one body the public entry point above and the `test-util`
    /// measurement in `assign/measured.rs` both run, so a suite asserting on
    /// the cost is asserting on the tally production publishes rather than on
    /// a second one written beside it.
    async fn select_recording(
        &self,
        runner_id: &Uuid7,
        now: UnixMillis,
    ) -> (Result<Option<Acquired>>, PollCost) {
        let mut cost = PollCost::default();
        let selected = self.select_counted(runner_id, now, &mut cost).await;
        // On EVERY exit path, including the one where the peek itself failed:
        // a poll that could not read the index is still a poll, and a total
        // that skipped it would make idle cost look lower than it is.
        producers::fleet::lease_polled(cost.candidates_scanned, cost.database_roundtrips);
        (selected, cost)
    }

    /// [`Leases::select`] without the recording, tallying what it cost.
    async fn select_counted(
        &self,
        runner_id: &Uuid7,
        now: UnixMillis,
        cost: &mut PollCost,
    ) -> Result<Option<Acquired>> {
        // One partition per poll, the next in the rotation: the read stays one
        // bounded round trip, and the partitions this poll did not visit are
        // the next polls' — whichever runner makes them.
        let partition = self.cursor().advance();
        let ready = self
            .ready()
            .peek(partition, MAX_READY_CANDIDATES_PER_POLL)
            .await
            .inspect_err(|error| warn_queue(EVENT_READY_PEEK_FAILED, runner_id, error))?;
        cost.candidates_scanned = u64::try_from(ready.len()).unwrap_or(u64::MAX);
        // The readiness depth this poll saw, published for the gauge that
        // reports it: the index is a network round trip, and a collection
        // callback cannot make one.
        producers::fleet::ready_depth_observed(cost.candidates_scanned);
        // The zero-Postgres path. Returning here is what makes idle cost scale
        // with runner count alone instead of runners × fleets.
        if ready.is_empty() {
            return Ok(None);
        }

        let ids: Vec<&str> = ready.iter().map(|entry| entry.fleet_id.as_str()).collect();
        cost.database_roundtrips += 1;
        for fleet_id in self.candidates(runner_id, &ids).await? {
            cost.database_roundtrips += 1;
            if let Some(acquired) = self.try_candidate(&fleet_id, runner_id, now).await? {
                return Ok(Some(acquired));
            }
        }
        Ok(None)
    }

    /// The eligible fleets among `ready`, in the query's own sticky order.
    ///
    /// The ordering must come from the statement and not from the peek, because
    /// sticky preference lives in its `ORDER BY`.
    async fn candidates(&self, runner_id: &Uuid7, ready: &[&str]) -> Result<Vec<Uuid7>> {
        let mut connection = self.pool().acquire().await?;
        let rows = sqlx::query(sql::lease::SELECT_READY_CANDIDATES)
            .bind(FLEET_STATUS_ACTIVE)
            .bind(runner_id.as_str())
            .bind(ready)
            .bind(i64::try_from(MAX_READY_CANDIDATES_PER_POLL).unwrap_or(i64::MAX))
            .fetch_all(&mut *connection)
            .await
            .map_err(query(CONTEXT_CANDIDATES))?;

        rows.iter()
            .map(|row| {
                let id: String = row.try_get(0).map_err(query(CONTEXT_CANDIDATES))?;
                Uuid7::parse(&id).map_err(crate::error::row_malformed("core.fleets", "id"))
            })
            .collect()
    }

    /// Claim one candidate and take its work, or answer `None` and move on.
    async fn try_candidate(
        &self,
        fleet_id: &Uuid7,
        runner_id: &Uuid7,
        now: UnixMillis,
    ) -> Result<Option<Acquired>> {
        // Recorded on the ONE exit that hands work out. A `None` — a lost
        // claim, an empty stream, a dropped entry — reaches nothing here,
        // because nothing started.
        self.try_candidate_unrecorded(fleet_id, runner_id, now)
            .await
            .inspect(|found| {
                if let Some(acquired) = found {
                    producers::fleet::run_started(started(acquired.kind));
                }
            })
    }

    /// [`Self::try_candidate`] without the recording, so the outcome exists
    /// before anything is said about it.
    async fn try_candidate_unrecorded(
        &self,
        fleet_id: &Uuid7,
        runner_id: &Uuid7,
        now: UnixMillis,
    ) -> Result<Option<Acquired>> {
        let Some(claimed) = self.claim(fleet_id, runner_id, now, LEASE_TTL_MS).await? else {
            // Taken by a live holder. No event was read, so nothing is orphaned.
            return Ok(None);
        };

        // A won claim over a lapsed holder means its lease is still `active`
        // and still names the work it never finished.
        if let Some(prior) = self.reclaim_prior_active(fleet_id, now).await? {
            let fleet = fleet_id.as_str();
            let runner = runner_id.as_str();
            let lease = prior.lease_id.as_str();
            let event_id = prior.event_id.as_str();
            let fence = claimed.fence.as_i64();
            tracing::debug!(
                event = EVENT_LEASE_RECLAIMED,
                fleet_id = fleet,
                runner_id = runner,
                lease_id = lease,
                agentsfleet_event_id = event_id,
                fencing_token = fence,
                "re-leasing a lapsed holder's event under a higher fence"
            );
            return from_reclaim(fleet_id, &claimed, prior).map(Some);
        }
        self.acquire_fresh(fleet_id, &claimed, now).await
    }

    /// Pull the next event for a claimed fleet: this consumer's own pending
    /// list first, then a new entry.
    ///
    /// Pending-first is safe precisely BECAUSE the claim was won: that proves no
    /// live lease exists, so a pending entry is a re-poll or a recovered strand
    /// rather than work somebody else is doing.
    async fn acquire_fresh(
        &self,
        fleet_id: &Uuid7,
        claimed: &crate::lease::affinity::Claimed,
        now: UnixMillis,
    ) -> Result<Option<Acquired>> {
        let streams = self.streams();
        let fleet = fleet_id.as_str();
        let Some(event) = self.read_fresh(fleet, &runner_consumer()).await? else {
            // Both reads answered, and both were empty — the only evidence this
            // code ever has that a fleet holds nothing deliverable. Free the
            // claim so the next event is not blocked behind it.
            self.release(fleet_id, claimed.fence, now).await?;
            return Ok(None);
        };
        match from_fresh(fleet_id, claimed, &event) {
            Ok(acquired) => Ok(Some(acquired)),
            Err(undecodable) => {
                drop_undecodable(&streams, fleet, &event.receipt, &undecodable).await;
                // Freed for the reason the empty arm frees it: this fleet holds
                // nothing this poll can lease. Holding the claim would cost a
                // full TTL of silence on a fleet whose next event may be fine.
                self.release(fleet_id, claimed.fence, now).await?;
                Ok(None)
            }
        }
    }
}

/// The label a granted lease's kind is counted under.
///
/// A total mapping: every kind a grant can carry has a start label, so a
/// grant cannot go uncounted — and a kind added without a label is a
/// compile error here rather than a series that never appears.
pub(crate) const fn started(kind: Kind) -> RunStart {
    match kind {
        Kind::Fresh => RunStart::Fresh,
        Kind::Reclaim => RunStart::Reclaimed,
    }
}

/// The stable consumer name this daemon reads under.
///
/// One name per process rather than per request: the pending list belongs to
/// the CONSUMER, so a name that changed per poll would strand every entry the
/// previous name had claimed.
///
/// `pub` because the reclaim sweeper claims stranded entries INTO this name and
/// the lease path reads OUT of it, and the two must be the same string. A
/// sweeper claiming into a name nothing reads would re-strand exactly the
/// entries it exists to rescue — the failure would look like the sweeper
/// working perfectly, since it would report claims every pass.
#[must_use]
pub fn runner_consumer() -> String {
    format!("agentsfleetd-{}", std::process::id())
}

#[cfg(feature = "test-util")]
pub mod measured;

#[cfg(all(test, feature = "test-util"))]
mod tests;
